de.dlr.sc.virsat.model/src/de/dlr/sc/virsat/model/dvlm/structural/util/StructuralElementInstanceHelper.java

```diff
@@ -108,4 +111,25 @@ public ICategoryAssignmentContainer getRoot() {
 		}
 		return superSeis;
 	}
+	
+	/**
+	 * Creates a new list which does not contain SEIs that are marked for deletion whose parents are also marked for deletion
+	 * for example, A contains B, and both are selected for deletion. In this case we need to remove B as it will be deleted anyway with A
+	 * @param seisToDelete Collection of SEIs that are selected for deletion
+	 * @return list of SEIs without children that will be deleted anyway
+	 */
+	public static ArrayList<StructuralElementInstance> cleanFromIndirectSelectedChildren(Collection<StructuralElementInstance> seisToDelete) {
+		Set<EObject> selectedToBeDeleted = new HashSet<>(seisToDelete);
+		ArrayList<StructuralElementInstance> seisToDeleteWithoutDuplicateChildren = new ArrayList<>();
+		for (StructuralElementInstance sei : seisToDelete) {
+			EObject parent = sei.eContainer();
+			while (parent instanceof StructuralElementInstance && !selectedToBeDeleted.contains(parent)) {

```

Comment:
Are we really interested in checking the type here or is it just a question of null/non-null? 
`while (parent != null && selectedToBeDeleted.contains(parent) { ... }` and `if (parent != null) { ... }` would be easier to read.

Reply:
No idea, I just moved and tested the already existing code. We have to investigate this.

Reply:
I think the way it is implemented is fine:
´´´
EObject parent = sei.eContainer();
while (parent instanceof StructuralElementInstance && !selectedToBeDeleted.contains(parent)) {
   parent = parent.eContainer();
}
´´´
parent is of type EObject. Even though our data model won't return anything else than a SEI or null, it is not practically checked when just using _null_ checks. I think _spotbugs_ may complain and we would have to introduce the type checking somewhere else. I can maybe add a comment to explain what is meant here.

Reply:
Looking deeper into the method, it looks like the algorithm is not correct. Therefore back to WIP.